### PR TITLE
problem_report: support reading zstd compressed values

### PR DESCRIPTION
**Pull request by bdrung**
`systemd-coredump` compresses the coredump with Zstandard. To support copying the coredump into a `ProblemReport` add support for reading Zstandard compressed values.

This merge request contains:
* https://github.com/canonical/apport/pull/255
* https://github.com/canonical/apport/pull/258
* https://github.com/canonical/apport/pull/259
* https://github.com/canonical/apport/pull/260
* https://github.com/canonical/apport/pull/261
* https://github.com/canonical/apport/pull/262
* https://github.com/canonical/apport/pull/269
* https://github.com/canonical/apport/pull/270

### CODE REVIEW

File: tests/unit/test_problem_report.py [hunk 1]
```diff
@@ -709,3 +709,73 @@ def test_get_on_disk_size(self) -> None:
...
+                First: this line is read
+                Second: this line is skipped
+                """
+            ).encode()
+        )
+        iterator = problem_report._EntryParser(report_file)
+        self.assertEqual(list(next(iterator)), [b"First: this line is read\n"])
+        next(iterator)
+        with self.assertRaises(StopIteration):
+            next(iterator)
```
Comment (schopin-pro) [inline tests/unit/test_problem_report.py:872]: issue: please add a test verifying that it skips empty lines :)

File: problem_report.py [hunk 2]
```diff
@@ -86,6 +86,54 @@ def __next__(self) -> Iterator[bytes]:
...
+            yield b"\n"
+        if line.endswith(b"\n"):
+            yield line[1:-1]
+            length += len(line) - 2
+        else:
+            yield line[1:]
+            length += len(line) - 1
+
+
+def _read_entry(entry: Iterator[bytes]) -> tuple[str, Iterator[bytes], bool]:
```
Comment (schopin-pro) [inline problem_report.py:132]: issue: documentation?

The final bool of the tuple isn't obvious.

File: problem_report.py [hunk 3]
```diff
@@ -199,77 +266,44 @@ def load(self, file, binary=True, key_filter=None):
...
+                remaining_keys.remove(key)
+                if len(remaining_keys) == 0:
+                    break
 
         self.old_keys = set(self.data.keys())
 
-    def extract_keys(self, file, bin_keys, directory):
-        # TODO: Split into smaller functions/methods
-        # pylint: disable=too-many-branches
+    def extract_keys(
```
Comment (schopin-pro) [inline problem_report.py:331]: thought: that name does not make any sense when reading the documentation.

**Review by schopin-pro — COMMENTED** [review]
Incomplete review, as I haven't looked at the last commit yet, but there's already some actionable feedback.

File: tests/unit/test_problem_report.py (hunk 1 above)
Comment (bdrung) [reply tests/unit/test_problem_report.py:872]: Empty lines are not allowed in the RFC822 format (but in deb822).

File: problem_report.py [hunk 4]
```diff
@@ -177,10 +225,29 @@ def add_tags(self, tags: Iterable[str]) -> None:
         new_tags = current_tags.union(tags)
         self["Tags"] = " ".join(sorted(new_tags))
 
-    def load(self, file, binary=True, key_filter=None):
-        # TODO: Split into smaller functions/methods
-        # pylint: disable=too-many-branches,too-many-nested-blocks
-        # pylint: disable=too-many-statements
+    def _decode_compressed_value(self, entry: Iterator[bytes]) -> Iterator[bytes]:
```
Comment (schopin-pro) [inline problem_report.py:255]: suggestion: a bit involved, but maybe move all compression knowledge into CompressedValue?

You'd need to transform CompressedValue to deal in iterators rather than scalar values, but I think it'd be worth it in terms of separation of concerns.

**Review by schopin-pro — COMMENTED** [review]
Some more actionable feedback, still haven't gotten to the last commit.

File: problem_report.py (hunk 2 above)
Comment (bdrung) [reply problem_report.py:132]: I renamed it to `_parse_entry` and added a docstring.

File: problem_report.py (hunk 3 above)
Comment (bdrung) [reply problem_report.py:367]: Yes, we should improve the wording there in a separate merge request. I opened https://github.com/canonical/apport/pull/271 to fix at least the docstring.

File: problem_report.py [hunk 5]
```diff
@@ -145,6 +198,25 @@ def get_on_disk_size(self) -> int:
         """
         return ((self.get_compressed_size() + 2) // 3) * 4
 
+    @staticmethod
+    def decode_compressed_value(entry: Iterator[bytes]) -> Iterator[bytes]:
```
[the remark this answers is not in the document]
Comment (bdrung) [reply problem_report.py:222]: I agree. I chose `decode_compressed_stream`.

File: problem_report.py [hunk 6]
```diff
@@ -31,6 +31,7 @@
 
 # magic number (0x1F 0x8B) and compression method (0x08 for DEFLATE)
 GZIP_HEADER_START = b"\037\213\010"
+ZSTANDARD_MAGIG_NUMBER = b"\x28\xB5\x2F\xFD"
```
[the remark this answers is not in the document]
Comment (bdrung) [reply problem_report.py:34]: Thanks. Fixed. (such kind of errors propagate due to the use of tab-completion).

File: .github/workflows/ci.yaml [hunk 7]
```diff
@@ -32,6 +32,10 @@ jobs:
           python3-distutils-extra python3-gi python3-launchpadlib
           python3-psutil python3-pytest python3-rpm python3-typeshed
           python3-yaml python3-requests
+      - name: Install optional dependencies
+        run: >
+          apt-get install --no-install-recommends --yes python3-zstandard
```
[the remark this answers is not in the document]
Comment (bdrung) [reply .github/workflows/ci.yaml:37]: python3-zstandard is not available on jammy. I added that as comment to `ci.yaml`.

File: problem_report.py [hunk 8]
```diff
@@ -204,6 +224,9 @@ def decode_compressed_stream(entry: Iterator[bytes]) -> Iterator[bytes]:
         block = next(entry, None)
         if block is None:
             return
+        if block.startswith(ZSTANDARD_MAGIC_NUMBER):
+            yield from _zstandard_decoder(entry, block)
```
Comment (schopin-pro) [inline problem_report.py:228]: praise: oooh, neat, `yield from`! Not something one sees often!

**Review by schopin-pro — APPROVED** [review]
LGTM!
